recipes/openai_textcat.py

```diff
@@ -0,0 +1,284 @@
+    segment=("Split sentences", "flag", "S", bool),
+    examples_path=("Examples file to help define the task", "option", "e", Path),
+    max_examples=("Max examples to include in prompt", "option", "n", int),
+    exclusive_classes=("Make the classification task exclusive", "flag", "E", bool),
+    verbose=("Print extra information to terminal", "flag", "v", bool),
+    # fmt: on
+)
+def textcat_openai_correct(
+    dataset: str,
+    input_path: Path,
+    prompt_path: Path,
+    labels: List[str] = None,
+    lang: str = "en",
+    model: str = "text-davinci-003",
+    batch_size: int = 10,
+    segment: bool = False,
+    examples_path: Optional[Path] = None,
+    max_examples: int = 2,
+    exclusive_classes: bool = False,
+    verbose: bool = False,
+):
+    api_key, api_org = get_api_credentials(model)
+    examples = read_prompt_examples(examples_path, example_class=TextCatPromptExample)
+    if labels is None:
+        msg.fail("textcat.teach requires at least one --label", exits=1)
+    nlp = spacy.blank(lang)
+
+    if segment:
+        nlp.add_pipe("sentencizer")
+
+    if not exclusive_classes and len(labels) == 1:
+        msg.warn(
+            "Binary classification should always be exclusive. Setting "
+            "`exclusive_classes` parameter to True"
+        )
+        exclusive_classes = True
+
+    # Create partial render of the template
+    template = Template(
+        load_template(prompt_path).render(
+            exclusive_classes=exclusive_classes, labels=labels, examples=examples
+        )
+    )
+
+    # Create OpenAISuggester with ChatGPT parameters
+    openai = OpenAISuggester(
+        response_parser=make_textcat_response_parser(labels=labels),
+        prompt_template=template,
+        labels=labels,
+        max_examples=max_examples,
+        segment=segment,
+        openai_api_org=api_org,
+        openai_api_key=api_key,
+        openai_n=1,
+        openai_model=model,
+        openai_retry_timeout_s=10,
+        openai_read_timeout_s=20,
+        openai_n_retries=10,
+        verbose=verbose,
+    )
+    for eg in examples:
+        openai.add_example(eg)
+
+    # Set up the stream
+    stream = list(srsly.read_jsonl(input_path))
+    stream = openai(tqdm.tqdm(stream), batch_size=batch_size, nlp=nlp)
+
+    # Set up the Prodigy UI
+    return {
+        "dataset": dataset,
+        "view_id": "blocks",
+        "stream": stream,
+        "update": openai.update,
+        "config": {
+            "labels": openai.labels,
+            "batch_size": batch_size,
+            "exclude_by": "input",
+            "choice_style": "single" if exclusive_classes else "multiple",
+            "blocks": [
+                {"view_id": "classification" if len(labels) == 1 else "choice"},
+                {"view_id": "html", "html_template": HTML_TEMPLATE},
+            ],
+            "show_flag": True,
+            "global_css": CSS_FILE_PATH.read_text(),
+        },
+    }
+
+
+@prodigy.recipe(
```

Comment:
I wonder ... can't we automatically read the `label` from the fetched examples? 

Here's the workflow I was trying. I first downloaded some news examples. 

```
python -m prodigy textcat.openai.fetch mini-news.jsonl templates/textcat_prompt.jinja2 news-out.jsonl --labels sports -F recipes/openai_textcat.py
```

They look like this: 

```
{
  "text": "Quest For The Thanksgiving Soda GASTONIA, NC -- Last year, the Jones Soda Co. rolled out, to much fanfare, the noxious-sounding product Turkey  amp; Gravy Soda. It flew off shelves faster than free antacids on the day after Thanksgiving, and ",
  "label": "sports",
  "openai": {
    "prompt": "\n\nFrom the text below, determine whether or not it contains a sports. If it is\na sports, answer \"accept.\" If it is not a sports, answer \"reject.\"\n\n\nYour answer should only be in the following format:\n\nanswer: <string>\nreason: <string>\n\nText:\n\"\"\"\nQuest For The Thanksgiving Soda GASTONIA, NC -- Last year, the Jones Soda Co. rolled out, to much fanfare, the noxious-sounding product Turkey  amp; Gravy Soda. It flew off shelves faster than free antacids on the day after Thanksgiving, and \n\"\"\"\n",
    "response": "\nAnswer: Reject\nReason: The text does not mention any sports."
  },
  "tokens": [...],
  "meta": {
    "answer": "REJECT",
    "reason": "The text does not mention any sports."
  },
  "answer": false,
  "_input_hash": 1874369427,
  "_task_hash": 677606533
}
```

Notice how there's a `label` key in there? It might be nice (and less error prone) if I didn't have to manually pass this label when calling `textcat.manual` here.

This is making me wonder ... does it make sense to have a `textcat.openai.manual` that can show me the `reject`/`accept` suggestion as well? As-is this doesn't get rendered in the standard textcat recipe. Such a recipe could also be used to filter specific, potentially rare, classes. 

It might be that I'm overlooking a workflow with another recipe though, so feel free to discuss that if I missed something.

Reply:
Hmm, that may be a good idea, but things we might want to think about:
- What if the labels in the input JSONL are not the full set of labels? (e.g., all the samples are `sports` but our task is to actually differentiate between `sports` and `hobby`)
- What if we decided to add another label in the annotation? (e.g., we've annotated everything as `sports` before, now we want to pass it again to Prodigy to differentiate between two more labels)

Perhaps providing the `labels` is the best way to do that? Is it more of a UI inconvenience? If that's the case then we can set `--labels` to `Optional` and then automatically fetch them if it's nothing is passed to `--labels`.

Reply:
> A user decided to add another label in the annotation (e.g., we've annotated everything as sports before, now we want to pass it again to Prodigy to differentiate between two more labels)

Ah, yeah you're right. Labels can totally change and the user should always be able to override them, otherwise we're indeed inconsistent with Prodigy. 

> I see, do you mean feeding things back to Prodigy but with the additional UI template of "OpenAI ChatGPT says: X"? 

Yes! Right now when I run `textcat.manual` on binary data then it doesn't show me the OpenAI suggestions that I would see in `textcat.openai.correct`, which would be nice to have around. 

> Can you explain the workflow more (especially the filtering rare classes part)?

The use-case was, for example, I'm interested in "sports" articles. So I'd like OpenAI to predict many examples and I only want to look at the examples where it predicted "sports". 

But I should've thought of that one more when I wrote it down. I figured a use-case here was to have OpenAI make predictions so that we might easily pluck the rare examples from it. But that's something the user should do by writing a custom script, not something a recipe should do.

Reply:
> I see, do you mean feeding things back to Prodigy but with the additional UI template of "OpenAI ChatGPT says: X"?

Come to think of it. It may also just be sufficient to show the parsed classes in the `meta` tag of the example. That way it'll still get rendered by `textcat.manual` and that should be sufficient to speed up annotations. 

![CleanShot 2023-01-24 at 12 50 22](https://user-images.githubusercontent.com/1019791/214284513-9158ce7d-027b-4ddf-afbf-c2f9d7f8c02e.png)

I guess this way you don't see the generated prompt, which is a downside, but it seems sufficient.

Reply:
Ah that makes sense! Will add that one in. Will probably change `OPINION` to `CHATGPT_SUGGESTION` (?)